fuzzers/main.py

```diff
@@ -0,0 +1,337 @@
+        def do_GET(self):
+            global FUZZERS
+            if self.path == '/pause':
+                signal_fuzzers(signal.SIGSTOP)
+                self.send_response(200)
+            elif self.path == '/resume':
+                signal_fuzzers(signal.SIGCONT)
+                self.send_response(200)
+            elif self.path == '/restart':
+                stop_fuzzers()
+                start_fuzzing(gcs)
+                self.send_response(200)
+            elif self.path == '/exit': # for development purposes only, production will be using systemctl stop
+                signal_fuzzers(signal.SIGKILL)
+                sys.exit(0)
+            else:
+                self.send_response(404)
+            self.end_headers()
+
+    with HTTPServer(("127.0.0.1", CMD_PORT), HTTPHandler) as httpd:
+        print(f"Serving command server on port {CMD_PORT}")
+        httpd.serve_forever()
+
+def start_fuzzing(gcs):
+    update_repo('master')
+    cfg = parse_config()
+    print(f"Parsed configuration: {cfg}")
+
+    # Checkout random branch
+    branch = random.choices(cfg['branch'], [b['weight'] for b in cfg['branch']])[0]
+    update_repo(branch['name'])
+
+    # TODO: Minimize the corpus
+    # TODO: Rsync the corpus from gcs more frequently, not just once per fuzzer restart
+    # TODO: Add corpus size gauge metric
+    # TODO: Add coverage metrics (will need to parse logs?)
+    # TODO: Add metrics about overhead (time to update corpus, to build fuzzer, etc.)
+    # TODO: Have one worktree per branch, this would allow both dependency build caching after a branch change and to run fuzzers on multiple branches from a single machine
+    #       The only drawback would be fuzz corpuses wouldn't be shared across processes on the same host on different branches, which is probably not too bad
+
+    # Start the fuzzers
+    for _ in range(NUM_FUZZERS):
+        start_a_fuzzer(branch, cfg, gcs)
+
+def reset_to_gcs(path, local, remote, log_file):
+    print(f"Resetting path {local / path} to GCS {remote}/{path}/")
+    log_file.write(f"Resetting path {local / path} to GCS {remote}/{path}/\n")
+    log_file.flush()
+    #utils.rmdirs(local / path)
+    os.makedirs(local / path, exist_ok = True)
+    #for blob in bucket.list_blobs(prefix = remote + '/' + path + '/'):
+    #    unprefixed_path = blob.name[len(remote) + 1:]
+    #    print(f"Downloading blob {blob.name} to {local / unprefixed_path}")
+    #    blob.download_to_filename(local / unprefixed_path)
+    subprocess.run(
+        ['gsutil', '-m', 'rsync', '-d', f"gs://{GCS_BUCKET}/{remote}/{path}/", local / path],
+        stdout = log_file,
+        stderr = subprocess.STDOUT,
+        check = False, # TODO: currently two concurrent gsutil can step on each other's feet, it still loads a corpus but it ends up with an error
+    )
+
+def auto_upload_to_gcs(exit_event, path, local, remote, bucket, log_file, ignore_deletions, on_uploaded_item, on_deleted_item):
```

Comment:
Can't we just sync it once the fuzzer finishes? I’m not entirely sure why we need to do it while the fuzzer is running. It’s not like we’re synchronising file systems of two different fuzzers. This will only upload files to GCS.

There could also be a case for perhaps setting up [NFS](https://cloud.google.com/architecture/partners/netapp-cloud-volumes/creating-nfs-volumes] or [SMB](https://cloud.google.com/architecture/partners/netapp-cloud-volumes/creating-smb-volumes) if it’s beneficial for fuzzers running on different machines to share corpus or artefacts.

Reply:
So the issue is that the fuzzer basically never finishes, so syncing only once the fuzzer finishes would mean that if we were to stop the process eg. for a reboot, we'd lose the last day of evolutions to the corpus. Another option would be to have a periodic sync, but it'd also hit another issue: it's hard to know, once the fuzzer evolution finished, which files have been deleted by the local fuzzer and which files have been added remotely by other fuzzers. (Sometimes the fuzzer replaces one element of the corpus by another one, in which case it's better to do the same on the GCS corpus)

As for setting up NFS or SMB, it's another option, that'd probably make things more efficient (in that they'd share the corpus more often than just once a day), but I think (though that'd probably deserve checking) that fuzzers are pretty IO-intensive (because they read, create and delete lots of small files), and thus the added latency would slow it down enough that better corpus sharing wouldn't be worth it.

Now TBH I'm not entirely sure about this, and it'd probably be worth adding the corpus size metrics, and then looking at the graphs with the two implementations to figure out which one is best.

(FWIW, one idea I had for further improvements was to have a hourly-or-so background task that'd rsync the local corpus to make it match the remote corpus, thus having the advantages of both world with fast local operations but still corpus shared with a reasonable sharing latency)

Reply:
If we delete or create new files while fuzzer is running will it even notice?

> So the issue is that the fuzzer basically never finishes, so syncing only once the fuzzer finishes would mean that if we were to stop the process eg. for a reboot, we'd lose the last day of evolutions to the corpus.

The disk is persistent so the data would still be there. Though it is true if we go without assumption that /datadrive is persistent, but even then reboot of the machine will be a rare event.

Reply:
> If we delete or create new files while fuzzer is running will it even notice?

Yes it will, the fuzzer runs inotify itself to automatically detect new elements in the corpus

> The disk is persistent so the data would still be there. Though it is true if we go without assumption that /datadrive is persistent, but even then reboot of the machine will be a rare event.

Right, so I guess the only valid argument there is the one about detecting which files were deleted by the fuzzer (because not deleting those files from the corpus would probably lead to a much too big corpus that takes way too long to load for no purpose) vs. which files were added to the gcs storage by other fuzzer instances on other machines (because they probably found new code paths)

Reply:
That would make using NFS or SMB complicated as they don’t really work with inotify.

I see why we might need it but I still hate it. ;) I wonder if there’s an existing tool which already handles synchronisation. rclone came to mind but I’m not sure if it has inotify support. 

If we share corpus and artifacts between commits, don’t we run the risk that a scenario will be tested in one branch and then not checked in another? It seems to me that we need to keep data for each commit separate.

Reply:
I think https://rclone.org/commands/rclone_mount/#vfs-file-caching be the best option if we only consider corpus sync, but still not work properly because we still need to do new artifact detection. It'd be the best option because it seems to me like with the full caching mode it should still be fast (by allowing for a slightly-outdated corpus) and automatically "upload"/"download" new artifacts. It probably wouldn't work with inotify, but upon checking the libfuzzer documentation I'm no longer sure whether it uses inotify or just periodically rescans the corpus by running ls.

However, we'd still need to inotify-watch on the artifacts folder in order to be able to notify only upon a new artifact being discovered (as opposed to a new fuzzer running the same artifact again and failing again), and judging from https://lwn.net/Articles/874000/ I'd guess it wouldn't be possible with the fuse-based rclone.

An option to alleviate this would be to have rclone mount create a directory, and then to use a 2-way-sync program that wouldn't rely on inotify but would keep its own list of files to sync with the actual corpus directory. Notwithstanding the question of whether such a program actually exists (maybe [inotify-sync](https://github.com/leeyiw/inotify-sync) would qualify? but it's small and abandoned, I only found it while searching for solutions), I feel like that'd be too much additional infrastructure to be worth the removal of the auto-uploader present in this code.

I unfortunately couldn't find anything that looked better − lsyncd looked like it might have fit the bill (it's basically an inotify-based rsync, though whether it'd behave correctly upon remote file addition would need checking), but it doesn't seem to support google cloud well, only supporting rsync and rsync-ssh syncing methods. (also, it looks like it would probably need patching in order to disable the initial sync, as we want to do the initial sync one way and then the follow-up changes sync the other way around)

-----

As for corpus and artifact sharing, the scenarios tested in one branch would still be checked in another: when starting up, the first thing a fuzzer does is go through the whole corpus and previous artifacts, run them again and report the crash if any such item generates a bug − and then it uses the result of this whole-corpus-scan in order to pick only a subset of the corpus that exercises most code paths to then mutate.

-----

So overall: given that we need to have the inotify probe in order to detect any new artifact anyway (because a fuzzer crash could be due to an existing artifact, and we wouldn't want to spam notifications in this case, as was happening in practice in my first local draft of this PR), I think the least bad option is to just keep the current auto-uploader, unless another program I couldn't find on the internet would exist to do just that